src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp

```diff
@@ -1596,24 +1611,1318 @@ struct npc_silverpine_forsaken_bat : public VehicleAI
+{
+    npc_silverpine_agatha_fenris_isle(Creature* creature) : ScriptedAI(creature), _healCD(false), _sceneStarted(false) { }
+
+    void JustAppeared() override
+    {
+        me->SetReactState(REACT_ASSIST);
+
+        me->RemoveNpcFlag(UNIT_NPC_FLAG_SPELLCLICK);
+
+        me->GetMotionMaster()->Clear();
+        me->GetMotionMaster()->MoveFollow(me->GetOwner(), 3.0f, float(M_PI / 2.0f));
+
+        _events.ScheduleEvent(EVENT_AGATHA_CHECK_PLAYER, 1s);
+    }
+
+    void IsSummonedBy(WorldObject* summoner) override
+    {
+        _playerGUID = summoner->GetGUID();
+    }
+
+    void SpellHit(WorldObject* /*caster*/, SpellInfo const* spellInfo) override
+    {
+        switch (spellInfo->Id)
+        {
+            case SPELL_AGATHA_BROADCAST:
+            {
+                if (_sceneStarted)
+                    return;
+
+                if (Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID))
+                    Talk(TALK_AGATHA_BROADCAST, player);
+                break;
+            }
+
+            case SPELL_GENERAL_TRIGGER_84114:
+                if (!_sceneStarted)
+                    SetEventNoEscape();
+                break;
+
+            case SPELL_GENERAL_TRIGGER_84107:
+                if (Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID))
+                    me->CastSpell(player, SPELL_ARMORE_CAMERA_1, true);
+                break;
+
+            case SPELL_ARMORE_CAMERA_4:
+                _events.ScheduleEvent(EVENT_FLEE_FROM_FENRIS + 1, 1s);
+                break;
+
+            case SPELL_GENERAL_TRIGGER_84079:
+                _events.ScheduleEvent(EVENT_FLEE_FROM_FENRIS + 2, 1s);
+                break;
+
+            case SPELL_DESPAWN_ALL_SUMMONS_AGATHA:
+                DoCastSelf(SPELL_MARK_MASTER_AS_DESUMMONED);
+                break;
+
+            default:
+                break;
+        }
+    }
+
+    void WaypointReached(uint32 waypointId, uint32 pathId) override
+    {
+        if (pathId == PATH_AGATHA_TO_FORSAKEN)
+        {
+            if (waypointId == WAYPOINT_SPEED_UP)
+                me->SetSpeed(MOVE_RUN, 15.880999f);
+            else if (waypointId == WAYPOINT_ARRIVED_TO_FORSAKEN)
+                _events.ScheduleEvent(EVENT_FLEE_FROM_FENRIS + 3, 100ms);
+        }
+    }
+
+    void Reset() override
+    {
+        _events.CancelEvent(EVENT_UNHOLY_SMITE);
+        _events.CancelEvent(EVENT_DOOMHOWL);
+
+        me->GetMotionMaster()->Clear();
+        me->GetMotionMaster()->MoveFollow(me->GetOwner(), 3.0f, float(M_PI / 2.0f));
+    }
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_UNHOLY_SMITE, 500ms);
+        _events.ScheduleEvent(EVENT_DOOMHOWL, 2s, 3s);
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        if (Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID))
+        {
+            if (!player->HasAura(SPELL_SUMMON_AGATHA_FENRIS) || !player->IsInWorld())
```

Comment:
can we move that into a generic spellscript `OnRemove` somehow? feels ugly to check for that every tick

Reply:
This spell has  `Despawn on Summoner Death` and `Despawn on Summoner Logout` SummonProperties

Reply:
However, those are NYI. Someone needs to implement those flags in order for me to be able to drop this workaround. I'll set a note as reference to it for now.